### PR TITLE
Fingerprinting fixes in client_golang

### PR DESCRIPTION
**Pull request by beorn7**
@juliusv These are the fixes we need in client_golang. prometheus/prometheus code that uses these fixes needs to be changed to not change the fingerprint used in storage.

(Separate PR imminent.)

### CODE REVIEW

File: model/metric.go [hunk 1]
```diff
@@ -26,14 +26,67 @@ var separator = []byte{0}
...
+	}
+	return true
 }
 
-// Before compares the fingerprints of both metrics.
+// Before compares the metrics, using the following criteria:
+//
+// If m has fewer labels than o, it is before o. If it has more, it is not.
+//
+// If the number of labels is the same, the subset of all label values is sorted
```
Comment (juliusv) [inline model/metric.go]: s/label values/labels names/?

File: model/signature.go [hunk 2]
```diff
@@ -46,30 +47,37 @@ func getHashAndBuf() *hashAndBuf {
...
 
 func putHashAndBuf(hb *hashAndBuf) {
+	hb.h.Reset()
+	hb.b.Reset()
 	hashAndBufPool.Put(hb)
 }
 
-// LabelsToSignature returns a unique signature (i.e., fingerprint) for a given
-// label set.
+// LabelsToSignature returns an quasi-unique signature (i.e., fingerprint) for a
```
Comment (juliusv) [inline model/signature.go]: s/an/a/

**Comment by juliusv** [comment]
:+1: besides nits

File: model/metric.go (hunk 1 above)
Comment (beorn7) [inline model/metric.go]: I screwed up here anyway, should be superset, not subset... will reword...

File: model/signature.go (hunk 2 above)
Comment (beorn7) [inline model/signature.go]: Done.

**Comment by juliusv** [comment]
:+1:

**Pull request merged**